src/peer_select.cc

```diff
@@ -754,8 +890,11 @@ PeerSelector::handlePingTimeout()
 {
     debugs(44, 3, url());
 
-    if (entry)
-        entry->ping_status = PING_DONE;
+    // do nothing if ping reply came while handlePingTimeout() was queued
```

Comment:
Which moment in time we assume as the 'timed out' point for a PeerSelector? I think that it should be the time of the preceding PeerSelectorPingMonitor::noteWaitOver()() call, because PeerSelector::handlePingTimeout() is called with a delay. In this case, we should update the 'timedout' statistics below at least.

Reply:
As we discussed, there is a difference between timeout from the PeerSelectorPingMonitor point of view and this code (and also high-level) point of view. From high-level point of view, the ping is considered timed out when we decide to react to the timeout. Similarly, from this method point of view, the ping is considered timed out when we decide to set ping.timedout and call selectMore().

Yes, a ping considered timed out by PeerSelectorPingMonitor might not be considered timed out by this code, but that is essentially inevitable due to the two code paths being in different asynchronous processing "threads". Moreover, this was already true before our changes due to the selectionAborted() exclusion, for arguably similar "something more important has happened earlier" reasons.